### ISSUE TITLE
Undo stops are created by cursor movement

### ISSUE THREAD

**Issue by axefrog** [microsoft/vscode#16475]
- VSCode Version: 1.7.2
- OS Version: Windows 10

I've just been updating my extension to make it so that automatic edits, made by the extension in between any two manual edits made by the user, are not included in the user's undo history and are simply removed as a batch as the user steps through the undo history made by their manual edits, so that the minor edits made by the extension are not given the same weight as a user's normal edits. Unfortunately, the `undoStopBefore` and `undoStopAfter` options seem to have no effect.

Steps to Reproduce:

1. Clone the simple extension _autotrim_ from https://github.com/axefrog/vscode-autotrim
2. Modify [line 36 of extension.ts](https://github.com/axefrog/vscode-autotrim/blob/master/src/extension.ts#L36), disabling undo stops so that the code matches the code block further below.
3. Set up a text file with five lines of text, and some trailing spaces on lines 2, 3 and 4.
4. Test the extension in debug mode by making a change to line 1 and then using the arrow keys to move the cursor to line 5 and make another edit there. The extension will trim lines 2, 3 and 4 as it moves past them on the way to line 5.
5. Try and undo the edits to lines 5 and 1.

**Bug:** I'd expect the first undo to remove the edit to line 5, and the next undo to reverse all of the automatic changes to lines 2, 3, and 4, as well as the manual edit made to line 1. Instead, undo stops are also created for each line that was automatically trimmed by the extension, requiring several additional undos before the manual edit to line 1 is undone.

The code for line 36 should be:

```ts
editor.edit(ed => ed.replace(line.range, text), {undoStopBefore: false, undoStopAfter: false});
```

**Comment by alexandrudima** [comment]
@axefrog `editor.edit` exists exactly for cases like this (i.e. bulk editing multiple regions in the file). I suggest that you execute the `for` loop inside the `editor.edit` call. i.e.:

```
editor.edit((editBuilder) => {

// for loop in here:
// i.e. multiple calls to editBuilder.replace

});
```

This will make it that this is one edit operation containing `N` edits instead of `N` edit operations containing 1 edit.

**Comment by axefrog** [comment]
@alexandrudima I'm not quite sure what you're suggesting. My example suggested multiple lines for clarity, but the real use case was simply that the extension was performing an edit and creating an undo stop where none was desired. I'm not sure I understand how your response addresses the original concern, i.e. that I make an edit, navigate around the document a bit, and then make another edit, after which I attempt an undo, only to have to step through one or more undo stops that are the result of the extension trimming lines between my manual edits. Note that multiple lines may have been trimmed, but they are not being trimmed as a single operation- they're being trimmed on an adhoc basis depending on user behaviour. I can't make multiple batched calls to perform those edits because they're not all happening at the same point in time. I also am unclear how your response addresses the fact that undo stops are created when the options specified that they shouldn't have been, irrespective of whether I made one edit or ten.

**Comment by alexandrudima** [comment]
@axefrog I'm sorry I misunderstood initially. 

**TL;DR** the API works, your edits do not create undo stops, but moving the cursor creates those undo stops, long explanation below

I have now tried your steps, and I can understand the intent of your extension: removing trailing whitespace on lines that have been "visited" by the cursor.

The root cause of this behaviour is that moving the cursor creates undo stops. Here I have logged in the undoStack manager what edit operations arrive to it and when undo stops are created:

![16475](https://cloud.githubusercontent.com/assets/5047891/21133501/61296caa-c119-11e6-9977-b8d105438454.gif)

I hope you notice how simply moving the cursor (without any edit operations) creates undo stops. Unfortunately, **this is by design**. The reason is simple: cursor moving is the single best hint of when to push undo stops in 90% of the cases. To further explain, consider the following steps:

### Typing does not push undo stops
* open an editor
* type `a`
* type `b`
* type `c`
* type `d`
* undo: notice how typing does not introduce undo stops and it is possible to undo `a`, `b`, `c` and `d` having been typed in a single undo operation.

![16475-ab](https://cloud.githubusercontent.com/assets/5047891/21133643/2fc4b8da-c11a-11e6-845c-f9a576295d48.gif)

### Cursor movement pushes undo stops
* open an editor
* type `a`
* type `b`
* press arrow left
* press arrow right
* type `c`
* type `d`
* undo: notice how the first undo removes `cd` and the second undo removes `ab`

![16475-ab-cd](https://cloud.githubusercontent.com/assets/5047891/21133700/7923b7a6-c11a-11e6-88f6-b6988cceb0d5.gif)

### In closing

This is expected and consistent with other editors I have tried. 

Notice how from a logical point of view, the edits are 100% the same in both cases. i.e:
* insert `a` at position (6, 0)
* after some time, insert `b` at position (6, 1)
* after some time, insert `c` at position (6, 2)
* after some time, insert `d` at position (6, 3)

The only difference between the two cases is that the cursor was moved, hence moving the cursor inserts undo stops. I welcome any suggestions for improving how this works (while keeping the current undo behaviour around typing, deleting, etc.), including in the form of a PR.

**Comment by alexandrudima** [comment]
@axefrog Having given it more thought, I find the intent of the extension cool, i.e. trim trailing whitespace on "visited" lines, but I also find it surprising that simply navigating a file leads to it being edited. i.e. if I open a file for reading and never intend to edit it, your extension will kick in and remove trailing whitespace.

What you could do to workaround the cursor pushing undo stops when it moves and to accomodate the "reading" case is that you could "buffer" the edits and send them off when you get a text document change event (you'll have to accomodate to ignore the document change events generated by your extension, i.e. don't enter an endless loop), but then when the end user creates an edit, your extension can react and trim the desired lines in one swift (i.e. bulk) edit containing `N` trim trailing whitespace edits.

**Comment by axefrog** [comment]
@alexandrudima Thanks a lot for the feedback and sorry for the slow reply. I'll take it on board and come back if I have anything further to add.
